src/gen_rules.ml

```diff
@@ -774,11 +774,43 @@ module Gen(P : Params) = struct
     | None -> []
     | Some name -> ["--cookie"; sprintf "library-name=%S" name]
 
+  (* Generate rules for the reason modules in [modules] and return a list
+     a new list of modules with only OCaml sources *)
+  let reason_rules ~dir ~modules =
+    let rule re =
+      let ml = Module.ocaml_of_reason re in
+      let refmt =
+        match Context.which ctx "refmt" with
+        | None -> die "refmt not found. Fix with $ opam install reason"
+        | Some refmt -> refmt in
+      let rule src target =
+        let src_path = Path.relative dir src in
+        Build.run (Dep refmt)
+          [ A "--print"
+          ; A "binary"
+          ; Dep src_path ]
+          ~stdout_to:(Path.relative dir target) in
+      let ml_rule = rule re.impl_fname ml.impl_fname in
+      match Option.both re.intf_fname ml.intf_fname with
+      | None -> [ml_rule]
+      | Some (s, t) -> [rule s t ; ml_rule]
+    in
+    String_map.map ~f:(fun (m : Module.t) ->
+      if m.reason then (
+        List.iter ~f:add_rule (rule m);
+        Module.ocaml_of_reason m
+      ) else (
+        m
+      )
+    ) modules
+
   (* Generate rules to build the .pp files and return a new module map where all filenames
      point to the .pp files *)
   let pped_modules ~dir ~dep_kind ~modules ~preprocess ~preprocessor_deps ~lib_name =
+    let modules = reason_rules ~dir ~modules in
     let preprocessor_deps = Dep_conf_interpret.dep_of_list ~dir preprocessor_deps in
     String_map.map modules ~f:(fun (m : Module.t) ->
+      assert (not m.reason);
```

Comment:
Instead of doing two successive `String_map.map`, we could replace this line by:

```ocaml
let m = ocaml_of_reason m in
```